### DISCUSSION TITLE
$self->ua vs. Mojo::UserAgent->new

### DISCUSSION THREAD

**Discussion by hussam-qasem** [mojolicious/mojo#2059] [Q&A] [answered]
I'm stating to fall in love with everything Mojolicious. I am converting an old application which uses:
* [`SOAP::Lite`](https://metacpan.org/pod/SOAP::Lite)
* [`REST::Client`](https://metacpan.org/pod/REST::Client)
* [`WWW::Mechanize`](https://metacpan.org/pod/WWW::Mechanize)

and I find [`Mojo::UserAgent`](https://docs.mojolicious.org/Mojo/UserAgent) is nearly a direct replacement for [`REST::Client`](https://metacpan.org/pod/REST::Client) and not-so-complex [`WWW::Mechanize`](https://metacpan.org/pod/WWW::Mechanize) use cases. Would you agree?

My other question is related to the re-use of `ua`. In my Controller, I'd like to call an external subroutine, e.g.: `MyApp::Network::Cool::getDetails($number)`.
* Should I pass `$self` to `getDetails($self, $number)` so I can re-use the `$self->ua`?
* Or should I simply do `my $ua  = Mojo::UserAgent->new;` inside `getDetails()`?

**Comment by kraih** [comment, marked as answer]
I would probably consider that a model, and store an instance of the class in a helper during startup:
```perl
$self->helper(cool => sub { state $cool = MyApp::Network::Cool->new($app->ua) });
```

**Reply by hussam-qasem** [reply]
Thank you! And do I need to `return $cool;`? or is that useless?

**Reply by kraih** [reply]
implicit return.

**Reply by hussam-qasem** [reply]
Thank you @kraih. And I'm sorry, I'm still struggling with syntax. In startup (decided to pass the whole `$self` so I can access `$self->config` too):
```perl
$self->helper(cool => sub { state $cool = MyApp::Network::Cool->new($self) });
```
In my Model:
```perl
package MyApp::Network::Cool;
use Mojo::Base -base, -signatures;

has 'app';
my $url = Mojo::URL->new('https://pie.dev/post');

sub get ($self, $number) {
  $url->query(number => $number);
  my $body = {'secret' => $self->config->{'secrets'};

  my $tx = $ua->post($url => json => $body);
  ...
}
```

However, `$self` has no access to `config`. Would you kindly guide me in the right direction?

Thanks again,

**Reply by flash548** [reply]
Probably should be something like:
`$self->helper(cool => sub { state $cool = MyApp::Network::Cool->new(app => $self) });`

then you can do
```
package MyApp::Network::Cool;
use Mojo::Base -base, -signatures;

has 'app';
has 'url' => sub { Mojo::URL->new('https://pie.dev/post') };

sub get ($self, $number) {
  $self->url->query(number => $number);
  my $body = {'secret' => $self->app->config->{'secrets'};

  my $tx = $self->app->ua->post($self->url => json => $body);
  ...
}
```

Disclaimer: I did not _try_ the above, just thought would show syntax.  

Regardless in your `MyApp::Network::Cool->new` constructor call, you need to assign `$self` object to something in your `MyApp::Network::Cool` class - as in my example I use the `app` property.  Also I converted the `url` property to the `has` style so that it can be more of an instance variable type and thus can be accessed with `$self->url` style.  Hope that helps some.

**Reply by flash548** [reply]
> Thank you! And do I need to `return $cool;`? or is that useless?

To add little more context in case you're newer to perl.  A perl sub's return (unless its explicitly `return`ed by a `return` statement), is the last evaluated expression/statement result, so as @kraih stated, in the example case the last expression was the state variable assignment to `$cool` so therefore `$cool` is implicitly returned.  But if you _wanted_ to be more verbose - yes `return $cool;` would also work.

**Reply by hussam-qasem** [reply]
Thank you @kraih and @flash548. This is starting to make sense, and works as expected too. I wasn't aware of the implicit return.
